libraries/botbuilder-azure/src/cosmosDbPartitionedStorage.ts

```diff
@@ -0,0 +1,255 @@
+/**
+ * @module botbuilder-azure
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Storage, StoreItems } from 'botbuilder';
+import { Container, CosmosClient, CosmosClientOptions } from '@azure/cosmos';
+import { CosmosDbKeyEscape } from './cosmosDbKeyEscape';
+import * as semaphore from 'semaphore';
+
+const _semaphore: semaphore.Semaphore = semaphore(1);
+
+/**
+ * Cosmos DB Partitioned Storage Options.
+ */
+export interface CosmosDbPartitionedStorageOptions {
+    /**
+     * The CosmosDB endpoint.
+     */
+    cosmosDbEndpoint: string;
+    /**
+     * The authentication key for Cosmos DB.
+     */
+    authKey: string;
+    /**
+     * The database identifier for Cosmos DB instance.
+     */
+    databaseId: string;
+    /**
+     * The container identifier.
+     */
+    containerId: string;
+    /**
+     * The options for the CosmosClient.
+     */
+    cosmosClientOptions?: CosmosClientOptions;
+    /**
+     * The throughput set when creating the Container. Defaults to 400.
+     */
+    containerThroughput?: number;
+}
+
+/**
+ * @private
+ * Internal data structure for storing items in a CosmosDB Collection.
+ */
+class DocumentStoreItem {
+    /**
+     * Gets the PartitionKey path to be used for this document type.
+     */
+    public static get partitionKeyPath(): string {
+        return '/id';
+    }
+    /**
+     * Gets or sets the sanitized Id/Key used as PrimaryKey.
+     */
+    public id: string
+    /**
+     * Gets or sets the un-sanitized Id/Key.
+     * 
+     */
+    public realId: string 
+    /**
+     * Gets or sets the persisted object.
+     */
+    public document: object
+    /**
+     * Gets or sets the ETag information for handling optimistic concurrency updates.
+     */
+    public eTag: string
+    /**
+     * Gets the PartitionKey value for the document.
+     */
+    public get partitionKey(): string {
+        return this.id;
```

Comment:
would this need to be `escapeKey(this.id)`?

Reply:
Good question. I don't *believe* it's necessary, since it looks like we escape it in all the necessary spots in the code. The dotnet side also doesn't have it. If you feel it's needed, I can add it and submit a PR for the dotnet side. Thoughts?

Reply:
seems odd that we'd set the id to escapeKey(id) and partition key to just id. That would imply that escapeKey isn't doing anything that the SDK wouldn't already do for you. Don't need to block this PR for it, but it seems like we should try removing escapeKey and seeing if we have any tests that fail.

Reply:
The [crazy keys test[(https://github.com/microsoft/botbuilder-js/pull/1282/files#diff-e509e24b560c2c2464b96c7c9b2e119bR125) fails when `escapeKey` calls are removed. I'm going to leave the code as-is, for this part.